File: internal/http/rpc.go
package http

import (
	"context"
	"errors"
	"log"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/zuccamia/career-planner/internal/applications"
	"github.com/zuccamia/career-planner/internal/companies"
	"github.com/zuccamia/career-planner/internal/discover"
	"github.com/zuccamia/career-planner/internal/i18n"
	"github.com/zuccamia/career-planner/internal/people"
	"github.com/zuccamia/career-planner/internal/profile"
	"github.com/zuccamia/career-planner/internal/sources/ats"
	"github.com/zuccamia/career-planner/internal/sources/scrape"
)

// Stateless RPC handlers: JSON in, JSON out; no persistence.

// ---- companies ----

// rpcLookupCompany wraps companies.Service.GuessCandidate for the browser client.
func (s *Server) rpcLookupCompany(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name           string `json:"name"`
		OutputLanguage string `json:"output_language"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	name := strings.TrimSpace(body.Name)
	if name == "" {
		writeErr(w, http.StatusBadRequest, "name is required")
		return
	}
	candidate, err := s.companies.GuessCandidate(r.Context(), name, body.OutputLanguage)
	if err != nil {
		// Return the fallback candidate the service produces on LLM failure,
		// with a warning in the body so the UI can surface it.
		log.Printf("rpc companies/lookup: %v", err)
		writeJSON(w, http.StatusOK, map[string]any{
			"candidate": candidate,
			"warning":   err.Error(),
		})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"candidate": candidate})
}

// rpcBuildDossier wraps companies.Service.Build for the browser client.
// Input: {official_name, website, ats_url, ats_provider}. Output: the Dossier
// struct (JSON-tagged). No DB access — the browser stores the result locally.
func (s *Server) rpcBuildDossier(w http.ResponseWriter, r *http.Request) {
	var body struct {
		OfficialName   string `json:"official_name"`
		Website        string `json:"website"`
		BlogURL        string `json:"blog_url"`
		ATSURL         string `json:"ats_url"`
		ATSProvider    string `json:"ats_provider"`
		OutputLanguage string `json:"output_language"`
		// Optional pre-scraped markdown for each URL. When the caller
		// (browser) has its own BYOK scraper active, it pre-scrapes and
		// passes content here. Empty fields cause the server to scrape
		// itself when a server-side scraper is configured. All non-required.
		WebsiteContent string `json:"website_content"`
		BlogContent    string `json:"blog_content"`
		CareersContent string `json:"careers_content"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	name := strings.TrimSpace(body.OfficialName)
	if name == "" {
		writeErr(w, http.StatusBadRequest, "official_name is required")
		return
	}
	website := strings.TrimSpace(body.Website)
	blogURL := strings.TrimSpace(body.BlogURL)
	atsURL := strings.TrimSpace(body.ATSURL)
	atsProvider := strings.TrimSpace(body.ATSProvider)
	pages := companies.Pages{
		Website: strings.TrimSpace(body.WebsiteContent),
		Blog:    strings.TrimSpace(body.BlogContent),
		Careers: strings.TrimSpace(body.CareersContent),
	}

	if s.serverScrapeAvailable(r.Context()) {
		if website != "" && atsURL == "" {
			if got, prov, err := ats.LookupATSURL(r.Context(), s.scrape, website); err != nil {
				log.Printf("dossier: ats lookup failed for %s: %v", website, err)
			} else if got != "" {
				atsURL = got
				if atsProvider == "" {
					atsProvider = prov
				}
			}
		}
		s.scrapeMissingContent(r.Context(), &pages, website, blogURL, atsURL)
	}

	out, err := s.companies.BuildDossier(r.Context(), companies.Company{
		OfficialName: name,
		Website:      website,
		BlogURL:      blogURL,
		ATSURL:       atsURL,
		ATSProvider:  atsProvider,
	}, body.OutputLanguage, pages)
	if err != nil {
		writeErr(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, out)
}

// scrapeMissingContent scrapes each URL in parallel into pages; skips prefilled slots.
func (s *Server) scrapeMissingContent(ctx context.Context, e *companies.Pages, website, blog, careers string) {
	var wg sync.WaitGroup
	fanout := func(label, url string, dst *string) {
		if url == "" || *dst != "" {
			return
		}
		wg.Add(1)
		go func() {
			defer wg.Done()
			s.scrapeInto(ctx, label, url, dst)
		}()
	}
	fanout("website", website, &e.Website)
	fanout("blog", blog, &e.Blog)
	fanout("careers", careers, &e.Careers)
	wg.Wait()
}

func (s *Server) scrapeInto(ctx context.Context, label, url string, dst *string) {
	res, err := s.scrape.Scrape(ctx, url, scrape.ScrapeOptions{
		Formats: []string{"markdown"}, OnlyMainContent: true,
	})
	if err != nil {
		log.Printf("dossier: %s scrape failed for %s: %v", label, url, err)
		return
	}
	*dst = res.Markdown
}

// ---- profile ----

// rpcGenerateBragTags wraps profile.Service.GenerateBragTags for the browser.
// Input: {"body":"..."}. Output: {"tags":[...]}.
func (s *Server) rpcGenerateBragTags(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Body           string `json:"body"`
		OutputLanguage string `json:"output_language"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	if strings.TrimSpace(body.Body) == "" {
		writeErr(w, http.StatusBadRequest, "body is required")
		return
	}
	tags, err := s.profile.GenerateBragTags(r.Context(), body.Body, body.OutputLanguage)
	if err != nil {
		writeServiceErr(w, r, "profile/generate-brag-tags", err)
		return
	}
	writeJSON(w, http.StatusOK, profile.BragTagResult{Tags: tags})
}

// Input: {"markdown":"...","output_language":"en|vi"}. Output: {"brags":[...]}.
func (s *Server) rpcImportBrags(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Markdown       string `json:"markdown"`
		OutputLanguage string `json:"output_language"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	if strings.TrimSpace(body.Markdown) == "" {
		writeErr(w, http.StatusBadRequest, "markdown is required")
		return
	}
	entries, err := s.profile.ImportBrags(r.Context(), body.Markdown, body.OutputLanguage)
	if err != nil {
		writeServiceErr(w, r, "profile/import-brags", err)
		return
	}
	writeJSON(w, http.StatusOK, profile.ImportBragsResult{Brags: entries})
}

// Input: {"markdown":"...","output_language":"en|vi"}. Output: profile.ImportedOverview.
func (s *Server) rpcImportOverview(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Markdown       string `json:"markdown"`
		OutputLanguage string `json:"output_language"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	if strings.TrimSpace(body.Markdown) == "" {
		writeErr(w, http.StatusBadRequest, "markdown is required")
		return
	}
	overview, err := s.profile.ImportOverview(r.Context(), body.Markdown, body.OutputLanguage)
	if err != nil {
		writeServiceErr(w, r, "profile/import-overview", err)
		return
	}
	writeJSON(w, http.StatusOK, overview)
}

// rpcImportResume wraps profile.Service.ImportResume.
// Input: {"source":"<markdown or typst>","output_language":"en|vi"}. Prompt
// is format-neutral so a single field suffices.
func (s *Server) rpcImportResume(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Source         string `json:"source"`
		OutputLanguage string `json:"output_language"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	source := strings.TrimSpace(body.Source)
	if source == "" {
		writeErr(w, http.StatusBadRequest, i18n.T(body.OutputLanguage, "profile.resumes.error.source_required"))
		return
	}
	resume, err := s.profile.ImportResume(r.Context(), source, body.OutputLanguage)
	if err != nil {
		writeServiceErr(w, r, "profile/import-resume", err)
		return
	}
	writeJSON(w, http.StatusOK, resume)
}

// ---- applications ----

// rpcAnalyzeRoleSignals analyzes JD + optional company dossier into a role brief
// the tailor pipeline uses as a rubric. Response: { brief: "<markdown>" }.
func (s *Server) rpcAnalyzeRoleSignals(w http.ResponseWriter, r *http.Request) {
	var in applications.AnalyzeRoleSignalsInput
	if !decodeJSON(r, w, &in) {
		return
	}
	out, err := s.applications.AnalyzeRoleSignals(r.Context(), in)
	if err != nil {
		writeServiceErr(w, r, "applications/analyze-role-signals", err)
		return
	}
	writeJSON(w, http.StatusOK, out)
}

// rpcTailor runs the composite tailor pipeline: rank per category, take
// top-N, draft the résumé.
func (s *Server) rpcTailor(w http.ResponseWriter, r *http.Request) {
	var in applications.TailorInput
	if !decodeJSON(r, w, &in) {
		return
	}
	out, err := s.applications.Tailor(r.Context(), in)
	if err != nil {
		writeServiceErr(w, r, "tailor", err)
		return
	}
	writeJSON(w, http.StatusOK, out)
}

// Wire shape shared by summarize/generate-message handlers.
type threadDetailPayload struct {
	Thread struct {
		PersonName  string `json:"person_name"`
		PersonNotes string `json:"person_notes"`
		Channel     string `json:"channel"`
		Subject     string `json:"subject"`
		Status      string `json:"status"`
		Summary     string `json:"summary"`
	} `json:"thread"`
	Entries []struct {
		Direction  string `json:"direction"`
		Content    string `json:"content"`
		OccurredAt string `json:"occurred_at"`
	} `json:"entries"`
	OutputLanguage string `json:"output_language"`
}

// toThreadDetail converts the wire payload into the domain type expected by
// buildThreadContext. occurred_at is parsed leniently: RFC3339 first, then
// SQLite's default "YYYY-MM-DD HH:MM:SS", then dropped to zero on failure.
func (p threadDetailPayload) toThreadDetail() people.ThreadDetail {
	detail := people.ThreadDetail{
		Thread: people.Thread{
			Person: people.Person{
				Name:  p.Thread.PersonName,
				Notes: p.Thread.PersonNotes,
			},
			Channel: p.Thread.Channel,
			Subject: p.Thread.Subject,
			Status:  p.Thread.Status,
			Summary: p.Thread.Summary,
		},
	}
	for _, e := range p.Entries {
		occurred, err := time.Parse(time.RFC3339, e.OccurredAt)
		if err != nil {
			occurred, _ = time.Parse("2006-01-02 15:04:05", e.OccurredAt)
		}
		detail.Entries = append(detail.Entries, people.ThreadEntry{
			Direction:  e.Direction,
			Content:    e.Content,
			OccurredAt: occurred,
		})
	}
	return detail
}

// ---- people ----

// rpcSummarizeThread wraps people.Service.SummarizeThreadContext.
// Input: thread + entries payload. Output: {"summary": "..."}. The browser
// owns persistence — this endpoint only runs the LLM prompt.
func (s *Server) rpcSummarizeThread(w http.ResponseWriter, r *http.Request) {
	var body threadDetailPayload
	if !decodeJSON(r, w, &body) {
		return
	}
	summary, err := s.people.SummarizeThreadContext(r.Context(), body.toThreadDetail(), body.OutputLanguage)
	if err != nil {
		log.Printf("rpc people/summarize-thread: %v", err)
		if errors.Is(err, people.ErrUnsafeGeneration) {
			writeErr(w, http.StatusBadRequest, i18n.T(body.OutputLanguage, "people.error.unsafe_summary"))
			return
		}
		writeErr(w, http.StatusBadGateway, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"summary": summary})
}

// rpcGenerateMessage wraps people.Service.GenerateMessageFromContext.
// Input: thread + entries + goal ("outreach" | "reply"). Output: {"message"}.
func (s *Server) rpcGenerateMessage(w http.ResponseWriter, r *http.Request) {
	var body struct {
		threadDetailPayload
		Goal string `json:"goal"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	message, err := s.people.GenerateMessageFromContext(r.Context(), body.toThreadDetail(), body.Goal, body.OutputLanguage)
	if err != nil {
		log.Printf("rpc people/generate-message: %v", err)
		if errors.Is(err, people.ErrUnsafeGeneration) {
			writeErr(w, http.StatusBadRequest, i18n.T(body.OutputLanguage, "people.error.unsafe_message"))
			return
		}
		writeErr(w, http.StatusBadGateway, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"message": message})
}

// rpcExtractJobDescription runs the full server-LLM JD flow: fetch (if raw
// is empty) → prompt → LLM → sanitize. Used by callers without BYOK LLM.
func (s *Server) rpcExtractJobDescription(w http.ResponseWriter, r *http.Request) {
	var body struct {
		CompanyName       string `json:"company_name"`
		RoleTitle         string `json:"role_title"`
		JobPostingURL     string `json:"job_posting_url"`
		JobDescriptionRaw string `json:"job_description_raw"`
		OutputLanguage    string `json:"output_language"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	structured, raw, err := s.applications.ExtractJD(r.Context(), applications.JDExtractionInput{
		CompanyName:       body.CompanyName,
		RoleTitle:         body.RoleTitle,
		JobPostingURL:     body.JobPostingURL,
		JobDescriptionRaw: body.JobDescriptionRaw,
		OutputLanguage:    body.OutputLanguage,
	})
	if err != nil {
		writeServiceErr(w, r, "applications/extract-job-description", err)
		return
	}
	// ExtractJD already logs the suspicious-input warning internally; just
	// surface it in the response so the browser can render it.
	writeJSON(w, http.StatusOK, struct {
		Structured        applications.JobDescriptionStructured `json:"structured"`
		JobDescriptionRaw string                                `json:"job_description_raw"`
		Warning           string                                `json:"warning,omitempty"`
	}{Structured: structured, JobDescriptionRaw: raw, Warning: applications.DetectSuspiciousJDInput(raw)})
}

// ---- scrape sidecars ----

// rpcDossierScrape fetches per-page markdown + discovers the ATS URL for a
// dossier. Called by BYOK-LLM browsers with no BYOK scraper; the browser
// runs the LLM itself.
func (s *Server) rpcDossierScrape(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Website        string `json:"website"`
		BlogURL        string `json:"blog_url"`
		ATSURL         string `json:"ats_url"`
		ATSProvider    string `json:"ats_provider"`
		WebsiteContent string `json:"website_content"`
		BlogContent    string `json:"blog_content"`
		CareersContent string `json:"careers_content"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	if !s.serverScrapeAvailable(r.Context()) {
		writeErr(w, http.StatusServiceUnavailable, "server scraper is not configured")
		return
	}
	website := strings.TrimSpace(body.Website)
	blogURL := strings.TrimSpace(body.BlogURL)
	atsURL := strings.TrimSpace(body.ATSURL)
	atsProvider := strings.TrimSpace(body.ATSProvider)
	pages := companies.Pages{
		Website: strings.TrimSpace(body.WebsiteContent),
		Blog:    strings.TrimSpace(body.BlogContent),
		Careers: strings.TrimSpace(body.CareersContent),
	}
	if website != "" && atsURL == "" {
		if got, prov, err := ats.LookupATSURL(r.Context(), s.scrape, website); err != nil {
			log.Printf("dossier: ats lookup failed for %s: %v", website, err)
		} else if got != "" {
			atsURL = got
			if atsProvider == "" {
				atsProvider = prov
			}
		}
	}
	s.scrapeMissingContent(r.Context(), &pages, website, blogURL, atsURL)
	writeJSON(w, http.StatusOK, struct {
		ATSURL         string `json:"ats_url"`
		ATSProvider    string `json:"ats_provider"`
		WebsiteContent string `json:"website_content"`
		BlogContent    string `json:"blog_content"`
		CareersContent string `json:"careers_content"`
	}{
		ATSURL: atsURL, ATSProvider: atsProvider,
		WebsiteContent: pages.Website,
		BlogContent:    pages.Blog,
		CareersContent: pages.Careers,
	})
}

// rpcApplicationScrape fetches a job posting via FetchPosting's ladder
// (known ATS → server scraper → generic HTTP) and returns enriched raw text
// + the ats.Posting struct. Called by BYOK-LLM browsers with no BYOK
// scraper; the browser runs the LLM itself.
func (s *Server) rpcApplicationScrape(w http.ResponseWriter, r *http.Request) {
	var body struct {
		JobPostingURL     string `json:"job_posting_url"`
		JobDescriptionRaw string `json:"job_description_raw"`
		OutputLanguage    string `json:"output_language"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	raw, posting, err := s.applications.FetchPosting(r.Context(), applications.PostingSource{
		URL:            body.JobPostingURL,
		Raw:            body.JobDescriptionRaw,
		OutputLanguage: body.OutputLanguage,
	})
	if err != nil {
		writeErr(w, http.StatusBadGateway, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, struct {
		EnrichedRaw string      `json:"enriched_raw"`
		Posting     ats.Posting `json:"posting"`
		Warning     string      `json:"warning,omitempty"`
	}{EnrichedRaw: raw, Posting: posting, Warning: applications.DetectSuspiciousJDInput(raw)})
}

// ---- discover ----

// POST /api/discover/run — runs the pipeline against the user's context,
// consuming any browser-precomputed inputs and returning ranked recs.

func (s *Server) rpcDiscoverRun(w http.ResponseWriter, r *http.Request) {
	var req discover.DiscoverRequest
	if !decodeJSON(r, w, &req) {
		return
	}
	resp, err := s.discover.Run(r.Context(), req)
	if err != nil {
		log.Printf("rpc discover run: %v", err)
		writeErr(w, http.StatusBadGateway, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, resp)
}

// rpcDiscoverSearch runs the search step only. Called by BYOK-LLM browsers
// that don't have BYOK search: expand + rank happen client-side, search
// borrows the server's SearXNG.
func (s *Server) rpcDiscoverSearch(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Query   discover.SearchQuery    `json:"query"`
		Profile discover.ProfileSummary `json:"profile"`
		Locale  string                  `json:"locale"`
	}
	if !decodeJSON(r, w, &body) {
		return
	}
	groups, err := s.discover.Search(r.Context(), body.Query, body.Profile, body.Locale)
	if err != nil {
		writeErr(w, http.StatusServiceUnavailable, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"groups": groups})
}

File: web/static/local/js/rpc.mjs
// Thin JSON-over-HTTP client for the Go RPC surface. Endpoints are stateless
// — the browser owns persistent data locally and calls Go only for shared
// business logic (LLM prompts + sanitization).
//
// When the user has BYOK configured (see storage/byok.mjs), llmCall assembles
// the prompt via /api/llm/prompts/:name, calls the user's provider directly
// from the browser, and posts the raw response to /api/llm/parse/:name for
// sanitization — the user's API key never touches the server.
//
// LLM callers accept an optional `onStep({name, status, error?})` callback
// so the UI can show a step-list progress indicator. Status: 'running' |
// 'done' | 'failed'. Step names: scrape, discover, prompt, generate, parse.

import { isByokActive, getByokConfig } from './storage/byok.mjs';
import { isScraperActive } from './storage/scraper.mjs';
import { scrapeWithCache, scrapeInParallel } from './scrape-with-cache.mjs';
import { discoverATSURL } from './discover-ats.mjs';
import { callOpenAICompatible, getServerLLMStatus } from './llm-client.mjs';
import { getServerScraperStatus } from './scrape-client.mjs';
import { currentLocale, t } from './i18n.mjs';

const DOSSIER_SCRAPE_TTL_SECONDS = 24 * 3600;
const JD_SCRAPE_TTL_SECONDS = 3600;

// Default onStep so LLM callers can invoke the callback unconditionally
// without guarding on undefined. Pages that want progress reporting pass a
// real callback (see ui/progress.mjs).
const noopStep = () => {};

// stepped wraps an async block with running/done/failed emissions on onStep.
// The wrapped fn's result is returned on success; on failure the error is
// re-thrown after the failed emit. hintKey (optional) renders as a sub-line
// under the step label — set it when the step encompasses hidden work the
// user might otherwise not know is happening.
const stepped = async (onStep, name, fn, hintKey) => {
  onStep({ name, status: 'running', hintKey });
  try {
    const out = await fn();
    onStep({ name, status: 'done' });
    return out;
  } catch (err) {
    onStep({ name, status: 'failed', error: err && (err.message || String(err)) });
    throw err;
  }
};

const post = async (url, body) => {
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  const text = await res.text();
  let payload = null;
  try { payload = text ? JSON.parse(text) : null; } catch { /* non-JSON error body */ }
  if (!res.ok) {
    const msg = payload?.error || text || `HTTP ${res.status}`;
    throw new Error(msg);
  }
  return payload;
};

const llmCall = async (name, input, serverPath, outputLanguage, onStep = noopStep) => {
  const lang = outputLanguage || currentLocale();
  const withLang = { ...input, output_language: lang };
  // build-dossier and extract-job-description are the two flows whose server
  // handlers may fetch web content when the caller didn't pre-scrape. When
  // the server has a scraper configured, tag the corresponding step with the
  // "includes web scraping" hint so users see why it's slower than usual.
  const mayServerScrape = name === 'build-dossier' || name === 'extract-job-description';
  const serverScrapeHint = mayServerScrape && (await getServerScraperStatus()).available
    ? 'progress.hint.server_scrape'
    : undefined;
  if (!(await isByokActive())) {
    // Fail fast when no server-side LLM is configured — otherwise the request
    // would 502/503 with a raw "llm client is not configured" string that
    // gives the user no path forward.
    const serverLLM = await getServerLLMStatus();
    if (!serverLLM.available) {
      throw new Error(t('settings.ai.error.no_llm_configured'));
    }
    // Server-side single round trip. Server does prompt-assembly + LLM +
    // parse internally, and — when SCRAPER_* is set — may also fetch web
    // content, all under the same "generate" step.
    return stepped(onStep, 'generate', () => post(serverPath, withLang), serverScrapeHint);
  }
  const cfg = await getByokConfig();
  // In BYOK mode the /api/llm/prompts/:name call is where the server-side
  // scrape happens (before prompt assembly). Hint the "prompt" step so the
  // user sees "includes web scraping" while it runs.
  const prompt = await stepped(onStep, 'prompt', () => post(`/api/llm/prompts/${name}`, withLang), serverScrapeHint);
  const raw = await stepped(onStep, 'generate', () => callOpenAICompatible(
    { system: prompt.system, user: prompt.user },
    cfg,
  ));
  const parseBody = { raw, input };
  if (name === 'extract-job-description') {
    parseBody.enriched_raw = prompt.enriched_raw;
    parseBody.posting = prompt.posting;
  }
  return stepped(onStep, 'parse', () => post(`/api/llm/parse/${name}`, parseBody));
};

// Ask the Go LLM pipeline to guess a canonical Company row from a typed name.
// Returns { candidate: {official_name, website, blog_url, ats_url, ats_provider, reasoning}, warning? }.
export const guessCompanyCandidate = (name, outputLanguage, onStep) =>
  llmCall('guess-candidate', { name }, '/api/companies/guess-candidate', outputLanguage, onStep);

// Known ATS host patterns. When a posting URL matches, we let the server's
// Greenhouse/Lever/Ashby providers extract structured fields the raw markdown
// would lose; otherwise the browser BYOK scraper handles it if active.
const KNOWN_ATS_HOSTS = [
  /^(job-)?boards\.greenhouse\.io$|\.greenhouse\.io$/i,
  /^jobs\.lever\.co$/i,
  /\.ashbyhq\.com$|^jobs\.ashbyhq\.com$/i,
];

const isKnownATSHost = (rawURL) => {
  try {
    const host = new URL(rawURL).hostname;
    return KNOWN_ATS_HOSTS.some(rx => rx.test(host));
  } catch { return false; }
};

// Extract structured facts from raw JD text. When job_description_raw is
// empty and job_posting_url is set, the source is fetched — client-side via
// the BYOK scraper for unknown ATS hosts, otherwise server-side.
export const extractJobDescription = async (input, outputLanguage, onStep = noopStep) => {
  const enriched = { ...input };
  const url = (input.job_posting_url || '').trim();
  const rawEmpty = !(input.job_description_raw || '').trim();
  if (rawEmpty && url && !isKnownATSHost(url) && (await isScraperActive())) {
    // Non-fatal — if the browser scrape fails we still let the server try.
    try {
      const md = await scrapeWithCache(url, {
        ttlSeconds: JD_SCRAPE_TTL_SECONDS,
        stepName: 'scrape',
        onStep,
      });
      if (md) enriched.job_description_raw = md;
    } catch (err) {
      console.warn('extractJobDescription: browser scrape failed, letting server try:', err);
    }
  }
  return llmCall('extract-job-description', enriched, '/api/applications/extract-job-description', outputLanguage, onStep);
};

// Build a dossier for the given company shape. When the browser has an
// active BYOK scraper, we scrape website + blog + ATS/careers page in
// parallel (cache-first) and — if ats_url is empty — try to discover it
// from the domain map. Scrape and discover failures are non-fatal: the
// dossier is built from whatever succeeded plus the structured fields.
export const buildDossier = async (company, outputLanguage, onStep = noopStep) => {
  const enriched = { ...company };
  const website = (company.website || '').trim();
  const blogURL = (company.blog_url || '').trim();

  if (await isScraperActive()) {
    // ATS discovery first (sequential) — if we discover an ATS URL here we
    // can also scrape it in the same parallel batch below. Best-effort.
    if (website && !enriched.ats_url) {
      try {
        const { atsURL, provider } = await stepped(onStep, 'discover', () => discoverATSURL(website));
        if (atsURL) {
          enriched.ats_url = atsURL;
          if (!enriched.ats_provider && provider) enriched.ats_provider = provider;
        }
      } catch (err) {
        console.warn('buildDossier: ATS discovery failed:', err);
      }
    }

    const scraped = await scrapeInParallel([
      { stepName: 'scrape_website', url: website,             key: 'website_content' },
      { stepName: 'scrape_blog',    url: blogURL,             key: 'blog_content'    },
      { stepName: 'scrape_careers', url: enriched.ats_url,    key: 'careers_content' },
    ], { ttlSeconds: DOSSIER_SCRAPE_TTL_SECONDS, onStep });
    Object.assign(enriched, scraped);
  }

  return llmCall('build-dossier', enriched, '/api/dossiers/build', outputLanguage, onStep);
};

// Generate suggested brag tags from the brag body only.
export const generateBragTags = (payload, outputLanguage, onStep) =>
  llmCall('generate-brag-tags', payload, '/api/profile/generate-brag-tags', outputLanguage, onStep);

// Ask the LLM to summarize one communication thread. Caller ships the full
// thread + entries context (the server is stateless for local-first data) and
// receives { summary }. Persisting the summary is the caller's job.
//
//   thread:  { person_name, person_notes, channel, subject, status, summary }
//   entries: [{ direction, content, occurred_at }] — newest first
export const summarizeThread = (payload, outputLanguage, onStep) =>
  llmCall('summarize-thread', payload, '/api/communications/summarize-thread', outputLanguage, onStep);

// Ask the LLM to draft outreach or a reply from a thread's context. Same
// payload shape as summarizeThread plus a `goal` of "outreach" or "reply".
// Returns { message }. The caller decides whether to save the draft.
export const generateMessage = (payload, outputLanguage, onStep) =>
  llmCall('generate-message', payload, '/api/communications/generate-message', outputLanguage, onStep);
